addOns/network/src/main/java/org/zaproxy/addon/network/internal/ratelimit/RateLimiterImpl.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2022 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.addon.network.internal.ratelimit;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.commons.httpclient.URIException;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.parosproxy.paros.network.HttpMessage;
+import org.zaproxy.zap.utils.Pair;
+
+/** Track rate limiting. */
+public class RateLimiterImpl implements RateLimiter {
+    private static final Logger LOGGER = LogManager.getLogger(RateLimiterImpl.class);
+    private final Map<RateLimiterKey, RateLimiterEntry> table = new ConcurrentHashMap<>();
+    private List<RateLimitParamRule> rules = new ArrayList<>();
+    private RateLimiterObserver observer;
+
+    @Override
+    public boolean throttle(HttpMessage message, int initiator) {
+        Pair<RateLimiterEntry, RateLimitParamRule> entryAndRule = getOrCreate(message, initiator);
+        if (entryAndRule == null) {
+            LOGGER.debug("Rate limit not requested for {}", message.getRequestHeader().getURI());
+            return false;
+        }
+        boolean result = false;
+        RateLimiterEntry limiterEntry = entryAndRule.first;
+        RateLimitParamRule rule = entryAndRule.second;
+        synchronized (limiterEntry) {
+            long millisToWait =
+                    (1000 / rule.getRequestsPerSecond())
+                            - (System.currentTimeMillis() - limiterEntry.getLastRequestTime());
+            if (millisToWait > 0) {
+                try {
+                    LOGGER.debug(
+                            "{}: sleeping for {} ms",
+                            message.getRequestHeader().getURI(),
+                            millisToWait);
+                    Thread.sleep(millisToWait);
+                    result = true;
+                } catch (InterruptedException e) {
+                    LOGGER.warn("Request sleep interrupted", e);
```

Comment:
This should not be a warn.

Reply:
The other places in network that catch InterruptedException are either logging at warn or throwing InterruptedIOException. Should I remove the log, mark as interrupted and throw InterruptedIOException to match the other code?

Reply:
Yes, this code will be called by others, it's up to them to decide how they want to handle the interruption (I'm fine with a `InterruptedIOException`), the case you mentioned is our own code so we need to handle the interruption ourselves (and logging is appropriate there).